pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,581 @@
+	ipClient, err := getIPClient(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create IP address client: %v", err)
+	}
+
+	ip, err := ipClient.Get(ctx, c.ResourceGroup, addrName, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IP %q: %s", addrName, err)
+	}
+
+	if ip.IPConfiguration == nil {
+		return nil, fmt.Errorf("IP %q has nil IPConfiguration", addrName)
+	}
+
+	ipAddresses := make([]string, 0)
+	if ip.IPConfiguration.PublicIPAddress != nil && ip.IPConfiguration.PublicIPAddress.IPAddress != nil {
+		ipAddresses = append(ipAddresses, *ip.IPConfiguration.PublicIPAddress.IPAddress)
+	}
+	if ip.IPConfiguration.PrivateIPAddress != nil {
+		ipAddresses = append(ipAddresses, *ip.IPConfiguration.PrivateIPAddress)
+	}
+
+	return ipAddresses, nil
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
+	return spec, false, nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, providerCfg, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	vmClient, err := getVMClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create VM client: %v", err)
+	}
+
+	osRef, err := getOSImageReference(providerCfg.OperatingSystem)
+	if err != nil {
+		return nil, err
+	}
+
+	// We genete a random SSH key, since Azure won't let us create a VM without an SSH key or a password
+	key, err := ssh.NewSSHKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ssh key: %v", err)
+	}
+
+	ifaceName := machine.Spec.Name + "-netiface"
+	iface, err := createNetworkInterface(context.TODO(), ifaceName, machine.UID, config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate main network interface: %v", err)
+	}
+
+	vmSpec := compute.VirtualMachine{
+		Location: &config.Location,
+		VirtualMachineProperties: &compute.VirtualMachineProperties{
+			HardwareProfile: &compute.HardwareProfile{VMSize: compute.VirtualMachineSizeTypes(config.VMSize)},
+			NetworkProfile: &compute.NetworkProfile{
+				NetworkInterfaces: &[]compute.NetworkInterfaceReference{
+					{
+						ID: iface.ID,
+						NetworkInterfaceReferenceProperties: &compute.NetworkInterfaceReferenceProperties{Primary: to.BoolPtr(true)},
+					},
+				},
+			},
+			OsProfile: &compute.OSProfile{
+				AdminUsername: to.StringPtr(adminUserName),
+				ComputerName:  &machine.Spec.Name,
+				LinuxConfiguration: &compute.LinuxConfiguration{
+					DisablePasswordAuthentication: to.BoolPtr(true),
+					SSH: &compute.SSHConfiguration{
+						PublicKeys: &[]compute.SSHPublicKey{
+							{
+								Path:    to.StringPtr(fmt.Sprintf("/home/%s/.ssh/authorized_keys", adminUserName)),
```

Comment:
wtf.

Reply:
Oh yeah. A value is required but currently only one value is supported...